pkg/controllers/machinesetsync/machineset_sync_controller.go

```diff
@@ -318,10 +320,53 @@ func (r *MachineSetSyncReconciler) reconcileMAPIMachineSetToCAPIMachineSet(ctx c
 		return result, fmt.Errorf("unable to ensure CAPI machine set: %w", err)
 	}
 
+	if result, err := r.deleteOutdatedCAPIInfraMachineTemplates(ctx, mapiMachineSet, newCAPIInfraMachineTemplate); err != nil {
+		return result, fmt.Errorf("unable to delete outdated Cluster API infrastructure machine templates: %w", err)
+	}
+
 	return ctrl.Result{}, r.applySynchronizedConditionWithPatch(ctx, mapiMachineSet, corev1.ConditionTrue,
 		consts.ReasonResourceSynchronized, messageSuccessfullySynchronizedMAPItoCAPI, &mapiMachineSet.Generation)
 }
 
+// deleteOutdatedCAPIInfraMachineTemplates deletes infra machine templates that have MAPI machine label of the current MachineSet and don't have the current computed hash.
+func (r *MachineSetSyncReconciler) deleteOutdatedCAPIInfraMachineTemplates(ctx context.Context, mapiMachineSet *machinev1beta1.MachineSet, newCAPIInfraMachineTemplate client.Object) (ctrl.Result, error) { //nolint:unparam
+	logger := log.FromContext(ctx)
+
+	machineSetLabelSelector := client.MatchingLabels(map[string]string{"machine.openshift.io/cluster-api-machineset": mapiMachineSet.Name})
+
+	infraMachineTemplates := &capiv1beta1.MachineSetList{}
```

Comment:
Why are we listing MachineSets? So we are deleting the MachineSet which removes the template and then it all gets regenerated? Is that not a little heavy handed?

Reply:
I completely missed that this is listing MachineSets, not infra machine templates. I can't work out how the tests pass!